### ISSUE TITLE
Why not combine the operation of with torch.no_grad() into the net.eval()?

### ISSUE THREAD

**Issue by songyuc** [pytorch/pytorch#31361]
## 🚀 Feature

I think it will be more convenient if PyTorch can combine the operation of `with torch.no_grad()` into the `net.eval()`.

## Motivation

When I was using PyTorch, I found it strange as I needed using the operations of `with torch.no_grad()` and the `net.eval()`, because I had thought the function `net.eval()` would complete the operation of closing the gradient passing in the parameters in my model, as its name *eval()*.
And I think, when using the model to infer the results, we tend to close the gradient updating.
So I am wondering whether there are some other reasons for you to separate the `with torch.no_grad()` from `net.eval()`.

## Pitch

That is, I just need to use `net.eval()` to change my model code into the state of validation, without the gradient updating, and do not need to use the code `with torch.no_grad()` extra.

## Additional context

I am also wondering that there may be some other reasons for you to separate the `with torch.no_grad()` from `net.eval()`, and I hope you can tell me the scenarios where the separating would not be suitable.

**Comment by albanD** [comment]
Hi,

I think the two have very different meaning:
- `net.eval()` is changing the behavior of the `nn.Module` to behave correctly for evaluation.
- `torch.no_grad()` is changing the behavior of the autograd to disable gradient computation.

You can use `net.eval()` during some part of training for example to freeze some batchnorm statistics and finetune your model. Or disable dropout during some phase of your training.
I am not saying here that these are common things but I have seen them being used in the wild.


That being said, maybe we can add an optional argument `net.eval(disable_grad=False)`.
So the current behavior will be kept, but you can give `True` for your forward to be automatically wrapped in `torch.no_grad()`?

**Comment by songyuc** [comment]
> Hi,
> 
> I think the two have very different meaning:
> 
> * `net.eval()` is changing the behavior of the `nn.Module` to behave correctly for evaluation.
> * `torch.no_grad()` is changing the behavior of the autograd to disable gradient computation.
> 
> You can use `net.eval()` during some part of training for example to freeze some batchnorm statistics and finetune your model. Or disable dropout during some phase of your training.
> I am not saying here that these are common things but I have seen them being used in the wild.
> 
> That being said, maybe we can add an optional argument `net.eval(disable_grad=False)`.
> So the current behavior will be kept, but you can give `True` for your forward to be automatically wrapped in `torch.no_grad()`?

Are you a developer of PyTorch?

**Comment by songyuc** [comment]
> Hi,
> 
> I think the two have very different meaning:
> 
> * `net.eval()` is changing the behavior of the `nn.Module` to behave correctly for evaluation.
> * `torch.no_grad()` is changing the behavior of the autograd to disable gradient computation.
> 
> You can use `net.eval()` during some part of training for example to freeze some batchnorm statistics and finetune your model. Or disable dropout during some phase of your training.
> I am not saying here that these are common things but I have seen them being used in the wild.
> 
> That being said, maybe we can add an optional argument `net.eval(disable_grad=False)`.
> So the current behavior will be kept, but you can give `True` for your forward to be automatically wrapped in `torch.no_grad()`?

In fact, I have figured out a case that when we want to do some operations like online learning, the separating the operation of  'torch.no_grad()' and the 'net.eval()' might be useful, as we might want to infer the results and finetune the model at the same time.

And I think your final suggestion is really wonderful! If there is an argument like 'net.eval(disable_grad=False)', as you said, it can keep the current behavior, and help us closing the gradient updating if we write the code, 'net.eval(disable_grad=True)'. And this change will also apply the case of online learning I mentioned above.

**Comment by albanD** [comment]
Yes.

The thing is that we currently have nice ways to write eval functions like:
```python
@torch.no_grad()
def eval(model, data):
  model.eval()
  # Rest of eval code
```

So it is not sure that such api would be useful. Is the only advantage of adding this that you can remove one line from your function definition?

**Comment by vadimkantorov** [comment]
would also be nice to have `with model.eval()` that would restore original model.training!

**Comment by songyuc** [comment]
> Yes.
> 
> The thing is that we currently have nice ways to write eval functions like:
> 
> ```python
> @torch.no_grad()
> def eval(model, data):
>   model.eval()
>   # Rest of eval code
> ```
> 
> So it is not sure that such api would be useful. Is the only advantage of adding this that you can remove one line from your function definition?

It is my honor to have a reply from you, the developer, and thanks for your contribution for PyTorch.
And I think this code,
```Python
@torch.no_grad()
def eval(model, data):
  model.eval()
  # Rest of eval code
```
may be useful, but I think there is not much difference like the code,
```Python
def eval(model, data):
  torch.set_grad_enabled(False)
  model.eval()
  # Rest of eval code
```
And I propose this idea, because, I think it will tend to be more logical and readable if you can combine `with torch.no_grad()` into `net.eval()`.
Besides, this combining will help me by not needing to add this extra line of `with torch.no_grad()`, which might cause extra calculation and result in slow training speed. Because I might always be worrying about forget adding the line.
Although it seems this combining might only reduce a line from my code.

From a perspective of a developer, I think closing gradient updating, BN and dropout are three common operations which would appear simultaneously in the validation phase, as shown in this demo,
[TRANSFER LEARNING FOR COMPUTER VISION TUTORIAL](https://pytorch.org/tutorials/beginner/transfer_learning_tutorial.html#training-the-model)
As a result, I think the best programming way is,
```Python
def eval(model, data):
  model.eval() # which is equal to model.eval(disable_grad=True) 
  # Rest of eval code
```
That is, the default value of disable_grad is True, instead of Fasle.

**Comment by fankib** [comment]
Please don't.

**Comment by albanD** [comment]
There is no plan to add the original feature request as the two have completely different goals.
Closing this to avoid any confusion.

Feel free to open a new issue if you want to propose a different feature/API.
